src/docs/perf/rendering/shader.md

```diff
@@ -0,0 +1,171 @@
+When a shader is first used, it needs to be compiled on the device.
+The compilation could cost up to a few hundred milliseconds
+whereas a smooth frame needs to be drawn within 16 milliseconds
+for a 60 fps (frame-per-second) display.
+Therefore, a compilation could cause tens of frames to be missed,
+and drop the fps from 60 to 6. This is _compilation jank_.
+After the compilation is complete, the animation should be smooth.
+
+Definitive evidence for the presence of shader compilation jank
+is to see `GrGLProgramBuilder::finalize` in the tracing.
+See the following screenshot for an example timeline tracing.
+
+![A tracing screenshot verifying jank]({% asset perf/render/tracing.png @path %}){:width="100%"}
+
+## What do we mean by "first run"?
+
+On Android, "first run" means that the user might see
+jank the first time opening the app. Subsequent opens
+should be fine.
+
+On iOS, "first run" means that the user might see
+jank the first time the user opens the app from scratch.
+
+## How to use SkSL warmup
+
+As of release 1.20, Flutter provides command line tools for app developers to
+collect shaders that might be needed for end-users in the SkSL
+(Skia Shader Language) format. The SkSL shaders can then be
+packaged into the app, and get warmed up (pre-compiled)
+when an end-user first opens the app, thereby reducing the compilation
+jank in later animations. Use the following instructions to collect
+and package the SkSL shaders:
+
+<ol markdown="1">
+<li markdown="1">Run the app with `--cache-sksl` turned on 
+    to capture shaders in SkSL:
+
+```sh
+flutter run --profile --cache-sksl
+```
+</li>
+
+<li markdown="1"> Play with the app to trigger as many animations
+    as needed; particularly those with compilation jank.
+</li>
+
+<li markdown="1"> Press `M` at the command line of `flutter run` to
+    write the captured SkSL shaders into a file named something like
+   `flutter_01.sksl.json`.
+</li>
+
+<li markdown="1"> Build the app with SkSL warm-up using the following,
+    as appropriate:
+
+Android:
+```sh
+flutter build apk --bundle-sksl-path flutter_01.sksl.json
+```
+
+iOS:
+```sh
+flutter build ios --bundle-sksl-path flutter_01.sksl.json
+```
+</li>
+
+<li markdown="1"> Test the newly built app.
+</li>
+</ol>
+
+Alternatively, you can write some integration tests to
+automate the first three steps using a single command.
+For example:
+
+```sh
+flutter drive --profile --cache-sksl --write-sksl-on-exit flutter_01.sksl.json -t test_driver/app.dart
+```
+
+With such [integration tests][], you can easily and reliably get the
+new SkSLs when the app code changes, or when Flutter upgrades.
+Such tests can also be used to verify the performance change
+before and after the SkSL warm-up. Even better, you can put
+those tests into a CI (continuous integration) system so the
+SkSLs are generated and tested automatically over the lifetime of an app.
+
+Take [Flutter Gallery][] as an example. The CI system is set up to
```

Comment:
Shall we specify this `Flutter Gallery` to be the old `Flutter Gallery` that's not in the Google play store?

Reply:
I didn't realize that!!!

Reply:
I didn't realize this, but I added a note.